dart/dynamics/ShapeFrame.h

```diff
@@ -0,0 +1,365 @@
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_DYNAMICS_SHAPEFRAME_H_
+#define DART_DYNAMICS_SHAPEFRAME_H_
+
+#include <Eigen/Dense>
+
+#include "dart/common/Signal.h"
+#include "dart/common/AddonWithVersion.h"
+#include "dart/common/SpecializedAddonManager.h"
+#include "dart/dynamics/FixedFrame.h"
+#include "dart/dynamics/TemplatedJacobianNode.h"
+#include "dart/dynamics/EllipsoidShape.h"
+
+namespace dart {
+namespace dynamics {
+
+class VisualAddon;
+class CollisionAddon;
+class DynamicsAddon;
+class ShapeFrame;
+
+namespace detail {
+
+struct VisualDataProperties
+{
+  /// Color for the primitive shape
+  Eigen::Vector4d mRGBA;
+
+  bool mUseDefaultColor;
+
+  /// True if this shape node should be kept from rendering
+  bool mHidden;
+
+  /// Constructor
+  VisualDataProperties(
+      const Eigen::Vector4d& color = Eigen::Vector4d(0.5, 0.5, 1.0, 1.0),
+      const bool hidden = false);
+
+  /// Destructor
+  virtual ~VisualDataProperties() = default;
+};
+
+struct CollisionDataProperties
+{
+  /// This object is collidable if true
+  bool mCollidable;
+
+  /// Constructor
+  CollisionDataProperties(const bool collidable = true);
+
+  /// Destructor
+  virtual ~CollisionDataProperties() = default;
+};
+
+struct DynamicsDataProperties
+{
+  /// Coefficient of friction
+  double mFrictionCoeff;
+
+  /// Coefficient of restitution
+  double mRestitutionCoeff;
+
+  /// Constructor
+  DynamicsDataProperties(const double frictionCoeff = 1.0,
+                         const double restitutionCoeff = 0.0);
+
+  /// Destructor
+  virtual ~DynamicsDataProperties() = default;
+};
+
+} // namespace detail
+
+class VisualAddon final :
```

Comment:
I would more prefer the pair of `VisualAddon` and `VisualAddonProperties` since `Data` sounds more general than `Addon` especially when we have `Addon` mechanics.